engines/macventure/script.cpp

```diff
@@ -0,0 +1,1213 @@
+* are too numerous to list here. Please refer to the COPYRIGHT
+* file distributed with this source distribution.
+*
+* This program is free software; you can redistribute it and/or
+* modify it under the terms of the GNU General Public License
+* as published by the Free Software Foundation; either version 2
+* of the License, or (at your option) any later version.
+
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+
+* You should have received a copy of the GNU General Public License
+* along with this program; if not, write to the Free Software
+* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*
+*/
+
+#include "common/system.h"
+
+#include "macventure/macventure.h"
+#include "macventure/script.h"
+#include "macventure/world.h"
+#include "macventure/container.h"
+
+namespace MacVenture {
+
+ScriptEngine::ScriptEngine(MacVentureEngine * engine, World * world) {
+	_engine = engine;
+	_world = world;
+	_scripts = new Container(_engine->getFilePath(kFilterPathID));
+}
+
+ScriptEngine::~ScriptEngine() {
+	if (_scripts)
+		delete _scripts;
+}
+
+bool ScriptEngine::runControl(ControlAction action, ObjID source, ObjID destination, Common::Point delta) {
+	EngineFrame frame;
+	frame.action = action;
+	frame.src = source;
+	frame.dest = destination;
+	frame.x = delta.x;
+	frame.y = delta.y;
+	frame.haltedInSaves = false;
+	frame.haltedInFirst = false;
+	frame.haltedInFamily = false;
+	_frames.push_back(frame);
+	debugC(3, kMVDebugScript, "SCRIPT: Stored frame %d, action: %d src: %d dest: %d point: (%d, %d)",
+		_frames.size() - 1, frame.action, frame.src, frame.dest, frame.x, frame.y);
+
+	return resume(true);
+}
+
+bool ScriptEngine::resume(bool execAll) {
+	debugC(3, kMVDebugScript, "SCRIPT: Resume");
+	while (_frames.size()) {
+		bool fail = execFrame(execAll);
+		if (fail) return true;
+	}
+	return false;
+}
+
+void ScriptEngine::reset() {
+	_frames.clear();
+}
+
+bool ScriptEngine::execFrame(bool execAll) {
+	bool doFirst = execAll;
+	bool doFamily = false;
+	bool fail;
+
+	EngineFrame *frame = &_frames.front();
+
+	// Do first dispatch script (script 0)
+	if (frame->haltedInFirst || doFirst) { // We were stuck or it's the first time
+		frame->haltedInFirst = false;
+		if (doFirst) { fail = loadScript(frame, 0); }
+		else { fail = resumeFunc(frame); }
+		if (fail) {
+			frame->haltedInFirst = true;
+			_engine->preparedToRun();
+			return true;
+		}
+		doFamily = true;
+		frame->familyIdx = 0;
+	}
+
+	// Do scripts in the family of player (ObjID 1)
+	if (frame->haltedInFamily || doFamily) { // We have to do the family or we were stuck here
+		frame->haltedInFamily = false;
+		Common::Array<ObjID> family = _world->getFamily(_world->getObjAttr(1, kAttrParentObject), false);
+		uint32 i = frame->familyIdx;
```

Comment:
Why i is not declared right in the for() statement. You're not using it out of that scope anyway.